### PR TITLE
Add type schemes.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: ast/src/analyzed/display.rs [hunk 1]
```diff
@@ -138,16 +159,20 @@ impl<T: Display> Display for FunctionValueDefinition<T> {
...
+            }) => format_outer_function(e, None, f),
+            FunctionValueDefinition::Expression(TypedExpression {
+                e,
+                type_scheme: Some(ty),
+            }) if *ty == Type::Col.into() => format_outer_function(e, None, f),
+            FunctionValueDefinition::Expression(TypedExpression {
+                e,
+                type_scheme: Some(ts),
+            }) => {
+                // TODO format type vars
```
Comment (chriseth) [inline ast/src/analyzed/display.rs:174]: todo

File: executor/src/constant_evaluator/mod.rs [hunk 2]
```diff
@@ -138,15 +141,18 @@ struct Symbols<'a, T> {
 }
 
 impl<'a, T: FieldElement> SymbolLookup<'a, T, FixedColumnRef<'a>> for Symbols<'a, T> {
-    fn lookup(&self, name: &str) -> Result<Value<'a, T, FixedColumnRef<'a>>, EvalError> {
+    fn lookup<'b>(&self, name: &str) -> Result<Value<'a, T, FixedColumnRef<'a>>, EvalError> {
         Ok(
             if let Some((name, _)) = self.computed_columns.get_key_value(name) {
+                // TODO it is actually not possible anymore to reference another fixed column,
```
Comment (chriseth) [inline executor/src/constant_evaluator/mod.rs:147]: TODO check if we can move this into the next PR

File: test_data/pil/book/declarations.pil [hunk 3]
```diff
@@ -3,15 +3,13 @@ namespace Main(16);
     // This defines a constant
     let rows = 2**16;
     // This defines a fixed column that contains the row number in each row.
-    let step = |i| i;
-    // This defines a copy of the column, also a fixed column because the type
-    // is explicitly specified.
-    let also_step: col = step;
+    // Only symbols whole type is "col" are considered fixed columns.
```
Comment (leonardoalt) [inline test_data/pil/book/declarations.pil:6]:
```suggestion
    // Only symbols whose type is "col" are considered fixed columns.
```

File: ast/src/parsed/display.rs [hunk 4]
```diff
@@ -649,6 +658,43 @@ impl<E: Display> Display for FunctionTypeName<E> {
...
 }
 
+pub fn format_type_scheme_around_name<E: Display>(
+    name: &str,
+    type_scheme: &Option<TypeScheme<E>>,
+) -> String {
+    if let Some(type_scheme) = type_scheme {
+        format!(
+            "{} {name}: {}",
+            type_scheme.type_vars_to_string(),
```
Comment (leonardoalt) [inline ast/src/parsed/display.rs:668]: is this the generics?

File: ast/src/parsed/display.rs [hunk 5]
```diff
@@ -649,6 +658,43 @@ impl<E: Display> Display for FunctionTypeName<E> {
...
+            self.0
+                .iter()
+                .map(|(v, b)| {
+                    format!(
+                        "{v}{}",
+                        if b.is_empty() {
+                            String::new()
+                        } else {
+                            format!(": {}", b.iter().join(" + "))
+                        }
```
Comment (leonardoalt) [inline ast/src/parsed/display.rs:690]: beautiful nested madness

File: ast/src/asm_analysis/display.rs [hunk 6]
```diff
@@ -47,13 +47,17 @@ impl<T: Display> Display for AnalysisASMFile<T> {
                 Item::Machine(machine) => {
                     write_indented_by(f, format!("machine {name}{machine}"), current_path.len())?;
                 }
-                Item::Expression(ExpressionWithTypeName { e, type_name }) => write_indented_by(
+                Item::Expression(ExpressionWithTypeScheme { e, type_scheme }) => write_indented_by(
                     f,
                     format!(
-                        "let {name}{} = {e};\n",
-                        type_name
+                        "let{} {name}{} = {e};\n",
```
Comment (leonardoalt) [inline ast/src/asm_analysis/display.rs:53]: This is done >=3 times

File: ast/src/analyzed/types.rs [hunk 7]
```diff
@@ -116,25 +198,86 @@ pub struct FunctionType {
...
+            0 => return self,
+            1 => {
+                let var = self.vars.vars().next().unwrap();
+                [(var.clone(), "T".to_string())].into()
+            }
+            _ => self
+                .vars
+                .vars()
+                .enumerate()
+                .map(|(i, v)| ((*v).clone(), format!("T{}", i + 1)))
```
Comment (leonardoalt) [inline ast/src/analyzed/types.rs:232]: why do we need these concrete names?

File: ast/src/analyzed/types.rs [hunk 8]
```diff
@@ -116,25 +198,86 @@ pub struct FunctionType {
...
+            ty,
+        }
+    }
+
+    /// Formats the type variables and bounds part.
+    pub fn type_vars_to_string(&self) -> String {
+        if self.vars.is_empty() {
+            String::new()
+        } else {
+            format!("<{}>", self.vars)
```
Comment (leonardoalt) [inline ast/src/analyzed/types.rs:258]: does the parser support this already?

File: parser/src/lib.rs [hunk 9]
```diff
@@ -72,6 +75,22 @@ pub fn parse_module<'a, T: FieldElement>(
...
+    let ctx = ParserContext::new(None, input);
+    TYPE_NAME_PARSER
+        .parse(&ctx, input)
+        .map_err(|err| handle_parse_error(err, None, input))
+}
+
+pub fn parse_type_var_bounds(input: &str) -> Result<TypeBounds, ParseError<'_>> {
+    let ctx = ParserContext::new(None, input);
+    TYPE_VAR_BOUNDS_PARSER
+        .parse::<GoldilocksField>(&ctx, input)
```
Comment (leonardoalt) [inline parser/src/lib.rs:90]: why Goldilocks?

File: pil-analyzer/src/statement_processor.rs [hunk 10]
```diff
@@ -437,20 +449,34 @@ where
     fn resolve_type_name(&self, mut n: TypeName<parsed::Expression<T>>) -> Result<Type, EvalError> {
         // Replace all expressions by number literals.
         for e in n.expressions_mut() {
-            let v = self.evaluate_expression(e.clone())?;
-            *e = parsed::Expression::Number(v);
+            let v = self.evaluate_expression_to_int(e.clone())?;
+            let v_u64 = v.to_u64().ok_or(EvalError::TypeError(format!(
```
Comment (leonardoalt) [inline pil-analyzer/src/statement_processor.rs:453]: why u64?

**Review by leonardoalt — COMMENTED** [review]
A few questions and small comments.

If fixed columns have to be explicitly declared as `col`, what happens in a case like this?

```
let x = |i| i;
col witness x = i + 2;
```

File: ast/src/analyzed/types.rs (hunk 7 above)
Comment (chriseth) [reply ast/src/analyzed/types.rs:232]: This is used for error reporting or if two type schemes are the same up to renaming of type variables.

File: ast/src/analyzed/types.rs (hunk 8 above)
Comment (chriseth) [reply ast/src/analyzed/types.rs:258]: Yes, the grammar symbol is "GenericTypedName"

File: ast/src/asm_analysis/display.rs (hunk 6 above)
Comment (chriseth) [reply ast/src/asm_analysis/display.rs:53]: Fixed in https://github.com/powdr-labs/powdr/pull/1083

File: ast/src/parsed/display.rs (hunk 4 above)
Comment (chriseth) [reply ast/src/parsed/display.rs:668]: Yes

File: parser/src/lib.rs (hunk 9 above)
Comment (chriseth) [reply parser/src/lib.rs:90]: True, this should be generic. We really need to change the grammar to use integers instead...

File: pil-analyzer/src/statement_processor.rs (hunk 10 above)
Comment (chriseth) [reply pil-analyzer/src/statement_processor.rs:453]: This is a number literal inside a type scheme, so it must be an array length. Will add a comment.

File: ast/src/parsed/display.rs (hunk 5 above)
Comment (leonardoalt) [reply ast/src/parsed/display.rs]: Nah not too important here

File: parser/src/lib.rs (hunk 9 above)
Comment (leonardoalt) [reply parser/src/lib.rs:92]: But what happens with Bn then?

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**